File: arrays-strings/01-two-sum.c
/*
 * Problem 01: Two Sum (LeetCode #1)
 * Topic: Arrays & Strings
 * Difficulty: Easy
 * Language: C
 */

#include <stdio.h>
#include <stdlib.h>

int* twoSum(int* nums, int numsSize, int target, int* returnSize) {
    int* result = (int*)malloc(2 * sizeof(int));
    *returnSize = 0;
    
    for (int i = 0; i < numsSize; i++) {
        for (int j = i + 1; j < numsSize; j++) {
            if (nums[i] + nums[j] == target) {
                result[0] = i;
                result[1] = j;
                *returnSize = 2;
                return result;
            }
        }
    }
    return result;
}

int main() {
    int nums[] = {2, 7, 11, 15};
    int target = 9;
    int returnSize;
    int* res = twoSum(nums, 4, target, &returnSize);
    
    if (returnSize == 2) {
        printf("Test 1: [%d, %d]\n", res[0], res[1]); // Expected: [0, 1]
    }
    free(res);
    return 0;
}

File: arrays-strings/01-two-sum.py
"""
Problem 01: Two Sum (LeetCode #1)
Topic: Arrays & Strings
Difficulty: Easy

Description:
Given an array of integers `nums` and an integer `target`, return indices of the 
two numbers such that they add up to `target`.

Time Complexity: O(n)
Space Complexity: O(n)
"""

from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        seen = {}
        for i, num in enumerate(nums):
            diff = target - num
            if diff in seen:
                return [seen[diff], i]
            seen[num] = i
        return []

if __name__ == "__main__":
    sol = Solution()
    print("Test 1:", sol.twoSum([2, 7, 11, 15], 9))  # Expected: [0, 1]
    print("Test 2:", sol.twoSum([3, 2, 4], 6))        # Expected: [1, 2]
    print("Test 3:", sol.twoSum([3, 3], 6))           # Expected: [0, 1]
